### ISSUE TITLE
Proposal: Consider adding warnings against using zfs native encryption and send/recv in production

### ISSUE THREAD

**Issue by AndrewJDR** [openzfs/openzfs-docs#494]
Among experienced zfs users and developers, it's conventional wisdom that zfs native encryption is not suitable for production usage, particularly when combined with snapshotting and zfs send/recv. There is a long standing data corruption bug:
https://github.com/openzfs/zfs/issues/12014

Additionally, if you join #zfs or #zfsonlinux on freenode and mention that you're having an issue with zfs native encryption, you'll be met with advice from developers that zfs native encryption is simply not reliable.

Should warnings be added to the sections of the documentation that mention native encryption that this combination of features (native encryption + send/recv) is known to be unsuitable for production usage? As-is, there don't appear to be any warnings, and it just seems inappropriate to guide new zfs users down a path toward potential data corruption.

**Comment by Matthew-Bradley** [comment]
If there is functionality around encryption that is known to cause corruption, there really ought to be an unavoidable warning in the software to act as a catch. Not everyone reads all relevant documentation before executing each command, and it seems like issues with encryption have been around long enough for warnings to be implemented (this isn't a freshly-discovered bug).

**Comment by rbrewer123** [comment]
@Matthew-Bradley that's a great idea.  Sounds to me like both the documentation and the software should warn against enabling ZFS-native encryption.  This is how the ZFS community can show respect and courtesy to users, especially considering that ZFS-native encryption has been causing real headaches and time-loss for real people for years now.

**Comment by wdoekes** [comment]
_(Disclaimer: I'm not nowledgeable about zfs internals. But I am an experienced user.)_

We've been running encryption on Ubuntu systems since it was available in the distro. _We have never had any corrupt data._

Until recently, the only problem we had was with Ubuntu/Jammy and a missing patch, which caused snapshots to be unmountable. The patch (or a send/recv loop) fixed that: https://bugs.launchpad.net/ubuntu/+source/zfs-linux/+bug/1987190

But, recently we _did_ observe the send/recv issues that people are talking about. This did make snapshots that were recently made unavailable/useless, although the data does appear to be there (according to zdb). This is tracked in #15474 but it very much looks like #12014.

https://github.com/openzfs/zfs/issues/15474#issuecomment-1827832617 (wdoekes, Nov 2023)
> **Reading with zdb -r does seem to work:** [...]

https://github.com/openzfs/zfs/issues/12014#issuecomment-860125077 (aerusso, May 2021)
> In my case, https://github.com/openzfs/zfs/issues/11688 (which you already reference), I've discovered that **rebooting "heals" the snapshot**

https://github.com/openzfs/zfs/issues/12014#issuecomment-865040511 (jgoerzen, May 2021)
> **After a reboot** but before a scrub, **the zfs send** you gave **executes fine.**

https://github.com/openzfs/zfs/issues/12014#issuecomment-1822764049 (J0riz, Nov 2023)
> **Rebooting the server** and running two scrubs afterwards **resolves the error.**

Usage pattern that appears to trigger this issue:

- a quick succession of zfs snapshot + zfs send (without --raw) of the same or different datasets

So. Yes, I would prefer that the bug gets fixed. But if you're willing to put up with maintenance of the occasional snapshot failures - which might never happen - then I think you should be fine.

**Comment by AndrewJDR** [comment]
I'm going to try to steer things back on topic to the idea of adding some warnings about this feature.

1. Even if these issues can always be worked around by rebooting and scrubbing twice (btw, you can find [reports](https://github.com/openzfs/zfs/issues/11688#issuecomment-1917425783) where this is not the case), the requirement of unexpected  reboots and multiple scrubs rules out many production use cases, and so it would still make some sort of warning in the documentation and/or tools justified. You can see @wohali toward the later half of https://github.com/openzfs/zfs/issues/11688#issuecomment-1916910483 articulating this clearly. Forced reboots and scrubs are something we can all probably tolerate on home/test lab servers, not on production systems that have hundreds/thousands of people depending upon it 24/7 and where scrubs can take hours or days.

2. Unfortunately, we also know that other sorts of issues exist with native encryption, because a zfs contributor has testbed reproduction cases that trigger: a) kernel panics b) corruption of encryption key data that requires special recovery that a non-zfs developer will probably not know how to perform. See my OP for more information on this.

**Comment by h-2** [comment]
> I'm going to try to steer things back on topic 

I do think it is on-topic to try to document as best as possible *when* these issues occur. It would strengthen the case for putting up a warning and help readers of such a warning to make an informed decision.

**Comment by Matthew-Bradley** [comment]
> I do think it is on-topic to try to document as best as possible _when_ these issues occur.
Agreed. There should be some clarity exactly \*what\* is being warned against.

Whether a specific warning/lockout is warranted when using certain features in combination with native encryption, depends on whether the problems can \*truly\* be narrowed to specific combinations. Right now is looks like the answer is No, with kernel panics and random data corruption in the mix. Even if that is disregarded, corruption with native encryption seems from this thread to impact multiple features across snapshots, send/receive, and scrubbing. In either case case top-level warnings against enabling native encryption in both the documentation and tools must be part of the solution. Some tooling friction against enabling it may also be warranted.

A corresponding issue should probably be set up in https://github.com/openzfs/zfs/issues or similar to track changes to the tooling.

The internet at large has already picked up on this (I first became aware of it from a phoronix article). The best thing the project can do is put strong safeguards in place to stop the flow of people being bitten. We want people to be absolutely certain that they can't do something dangerous without _at least_ running into a warning or error.

**Comment by AndrewJDR** [comment]
Absolutely agreed that having as much clarity as we can achieve is a good thing. If anyone has suggestions on tweaks for the warning message based on what you've learned, please chime in -- the draft is in the OP. I've tried to clarify it as much as I can, based on what I've been able to learn from the publicly available information. If someone wants to work in the testbed results from @rincebrain, that seems fine as well. I thought about doing it, but struggled with how to phrase it.

Personally, I think it's important not to make the message "too scary", because that can provide folks an opening to muddy the waters with comments like "Well, I've been using it fine for years!", which while not untrue, doesn't really help the many people that try it and run into the issues. This is why the current draft of the warning mentions that many have been able to use it without issue.

**Comment by robszy** [comment]
I think the only way is to fix issues if we have encryption merged so leaders agreeded to support it like there is support when there is a bug in code without encryption.

Otherwise there should be warning encryption is EXPERIMENTAL - no one supports it. Use if you are prepared to lose your data.

**Comment by grahamperrin** [comment]
- https://github.com/openzfs/zfs/pull/16745

**Comment by grahamperrin** [comment]
<https://github.com/openzfs/openzfs-docs/issues/494#issuecomment-1946981043>

> … The internet at large has already picked up on this

FreeBSD Discord drew attention to <https://old.reddit.com/r/zfs/comments/1aowvuj/-/kq4f13l/> (2024-02-12) in _PSA: ZFS has a data corruption bug when using native encryption and send/recv_.

> … aware of it from a phoronix article). …

Reference: [OpenZFS Native Encryption Use Raises Data Corruption Concerns - Phoronix](https://www.phoronix.com/news/OpenZFS-Encrypt-Corrupt) – I assume that this article triggered the same-day post in /r/zfs.

**Comment by grahamperrin** [comment]
For anyone who's curious, the expectation of a _Bugs_ section at the foot of a man page arises from: 

- <https://man.freebsd.org/cgi/man.cgi?query=groff_mdoc&sektion=7&manpath=freebsd-ports#A_man_page_template> – a _man page template_

– and for **usage examples**, groff_mdoc(7) above is one of two pages to which we're directed from the [Manual pages](https://docs.freebsd.org/en/books/fdp-primer/manual-pages/) chapter of the _FreeBSD Documentation Project Primer for New Contributors_ a.k.a. the _FDP Primer_. 

----

If the notion of a Bugs section is contentious, in an OpenZFS context: 

- sorry
- probably best discussed elsewhere – re: <https://openzfs.org/wiki/Newcomers#I_have_a_question!>, I'll occasionally make myself available in IRC (bridged with Slack).

HTH

**Comment by AndrewJDR** [comment]
The native encryption + snapshot corruption issue can be consistently reproduced using the following [repro scripts](https://github.com/HankB/provoke_ZFS_corruption/tree/main) and [repro steps](https://hankb.github.io/provoke_ZFS_corruption/#2025-02-11-test-instructions).

The author of the reproducer also bisected it with the following conclusion:
https://hankb.github.io/provoke_ZFS_corruption/#2025-04-08-bisect-conclusion

Unfortunately, the particular commit where the issue started happening (https://github.com/openzfs/zfs/commit/30af21b02569ac192f52ce6e6511015f8a8d5729) is quite large, and it doesn't bring as much immediate clarity to the cause or a fix as one might hope for. The issue is being tracked here: https://github.com/openzfs/zfs/issues/12014

**Comment by alecStewart1** [comment]
@AndrewJDR you're right. Misread the reply I quoted...got a little too excited.

**Comment by grahamperrin** [comment]
@AndrewJDR thanks, your first mentioned issue does seem to be the one to watch: 

- https://github.com/openzfs/zfs/issues/12014

– a commit to the `zfs-2.2.8-staging` branch, and so on.

**Comment by AndrewJDR** [comment]
I've updated my OP. This is all very exciting, however I do want to quote [this exchange from the PR that seemingly fixed the issues](https://github.com/openzfs/zfs/pull/17340#issuecomment-2899706843):

> > Don't get me wrong; this PR is obviously right for what it is, and I won't be surprised if it closes a locking gap somewhere that could be hit if you're unlucky, but I'd like those dots connected so I can feel comfortable saying yes, this definitely sorts it out. Especially since the forums are starting to pick up on it and the high-fives and cheers are making me a bit uncomfortable. (robn)

> @robn I don't think there is still a proper understanding of what's going on. At least I haven't seen and don't have one. (amotin)

Basically no one understands precisely **_why_** these particular code defects were causing these particular issues, so it's probably best not to count our eggs just yet. It might be good to leave this issue open until the patches have been out in the wild for a few months.

But fingers crossed!

**Comment by omentic** [comment]
> Personally, I think it's important not to make the message "too scary", because that can provide folks an opening to muddy the waters with comments like "Well, I've been using it fine for years!", which while not untrue, doesn't really help the many people that try it and run into the issues. This is why the current draft of the warning mentions that many have been able to use it without issue.

I think it should be more heavily emphasized this is only generally considered an issue with non-raw sends. I use raw sends for encrypted snapshots because that's just the most straightforward and easiest thing to do, and most everything around zfs native encryption + send/recv that I've seen also takes this approach. I got very scared reading the start of this issue.

> In some cases, the issue can be avoided entirely by using a "raw" zfs send instead of an unencrypted zfs send, but it is not known with absolute certainty if this is always successful in avoiding the issue.

Are there any known cases of zfs native encryption + a **raw** send/recv leading to corruption?

**Comment by scineram** [comment]
There is no non-raw send issue.
The raw send issue is exactly described in the blog @ophilli linked. Maybe it could be possible to detect that failure mode on the send or receive side and then abort the replication. But some of the devs would have to chime in on that.

**Comment by owlshrimp** [comment]
> > Personally, I think it's important not to make the message "too scary", because that can provide folks an opening to muddy the waters with comments like "Well, I've been using it fine for years!", which while not untrue, doesn't really help the many people that try it and run into the issues. This is why the current draft of the warning mentions that many have been able to use it without issue.
> 
> I think it should be more heavily emphasized this is only generally considered an issue with non-raw sends. I use raw sends for encrypted snapshots because that's just the most straightforward and easiest thing to do, and most everything around zfs native encryption + send/recv that I've seen also takes this approach. I got very scared reading the start of this issue.
> 
> > In some cases, the issue can be avoided entirely by using a "raw" zfs send instead of an unencrypted zfs send, but it is not known with absolute certainty if this is always successful in avoiding the issue.
> 
> Are there any known cases of zfs native encryption + a **raw** send/recv leading to corruption?

afaik this has already been completely fixed as of earlier this year in may.

**Comment by owlshrimp** [comment]
Considering this issue was fixed months ago, this proposed warning likely serves no purpose.

**Comment by AndrewJDR** [comment]
Yes, I'm going to close this issue, as the root cause seems to have been fixed. If any evidence pops up showing there are still issues, please detail those, and this could be reopened.
